From 8d3df0c886f4393fa8e1dacd2a3532f8d913ad97 Mon Sep 17 00:00:00 2001
From: Sriram R <quic_srirrama@quicinc.com>
Date: Wed, 23 Nov 2022 14:26:24 +0530
Subject: [PATCH] mac80211: use link conf and link sta

Use link conf and link sta for link specific operations.
The link conf is replaced in start ap.
Link sta is updated with NSS during link station add.

Also, moved the condition for link bss address validation
to differentiate link bss address from ML address.

Signed-off-by: Sriram R <quic_srirrama@quicinc.com>
---
 net/mac80211/cfg.c | 22 +++++++++++-----------
 1 file changed, 11 insertions(+), 11 deletions(-)

--- a/net/mac80211/cfg.c
+++ b/net/mac80211/cfg.c
@@ -1333,40 +1333,40 @@ static int ieee80211_start_ap(struct wip
 	}
 
 	if (params->he_cap) {
-		memcpy(&sdata->vif.bss_conf.he_cap_elem, params->he_cap,
+		memcpy(&link_conf->he_cap_elem, params->he_cap,
 		       sizeof(*params->he_cap));
-		sdata->vif.bss_conf.he_su_beamformer =
+		link_conf->he_su_beamformer =
 			params->he_cap->phy_cap_info[3] &
 			IEEE80211_HE_PHY_CAP3_SU_BEAMFORMER;
-		sdata->vif.bss_conf.he_su_beamformee =
+		link_conf->he_su_beamformee =
 			params->he_cap->phy_cap_info[4] &
 			IEEE80211_HE_PHY_CAP4_SU_BEAMFORMEE;
-		sdata->vif.bss_conf.he_mu_beamformer =
+		link_conf->he_mu_beamformer =
 			params->he_cap->phy_cap_info[4] &
 			IEEE80211_HE_PHY_CAP4_MU_BEAMFORMER;
-		sdata->vif.bss_conf.he_full_ul_mumimo =
+		link_conf->he_full_ul_mumimo =
 			params->he_cap->phy_cap_info[2] &
 			IEEE80211_HE_PHY_CAP2_UL_MU_FULL_MU_MIMO;
 	}
 
 	if (params->eht_cap) {
-		if (!sdata->vif.bss_conf.he_support)
+		if (!link_conf->he_support)
 			return -EOPNOTSUPP;
 
-		sdata->vif.bss_conf.eht_support = true;
-		sdata->vif.bss_conf.eht_su_beamformer =
+		link_conf->eht_support = true;
+		link_conf->eht_su_beamformer =
 				params->eht_cap->fixed.phy_cap_info[0] &
 				IEEE80211_EHT_PHY_CAP0_SU_BEAMFORMER;
-		sdata->vif.bss_conf.eht_su_beamformee =
+		link_conf->eht_su_beamformee =
 				params->eht_cap->fixed.phy_cap_info[0] &
 				IEEE80211_EHT_PHY_CAP0_SU_BEAMFORMEE;
-		sdata->vif.bss_conf.eht_mu_beamformer =
+		link_conf->eht_mu_beamformer =
 				params->eht_cap->fixed.phy_cap_info[7] &
 				IEEE80211_EHT_PHY_CAP7_MU_BEAMFORMER_80MHZ;
-		sdata->vif.bss_conf.eht_80mhz_full_bw_ul_mumimo =
+		link_conf->eht_80mhz_full_bw_ul_mumimo =
 				params->eht_cap->fixed.phy_cap_info[7] &
 				IEEE80211_EHT_PHY_CAP7_NON_OFDMA_UL_MU_MIMO_80MHZ;
-		ieee80211_eht_op_ie_to_bss_conf(&sdata->vif, params->eht_oper);
+		ieee80211_eht_op_ie_to_bss_conf(link_conf, params->eht_oper);
 	}
 
 	mutex_lock(&local->mtx);
@@ -1893,6 +1893,15 @@ static int sta_link_apply_parameters(str
 			link_sta->pub->ru_punct_bitmap = params->ru_punct_bitmap;
 	}
 
+	/* update nss if not done already for the link sta,
+	 * for the deflink the nss will be updated once the assoc
+	 * flag is set through set station. For the non-deflink sta
+	 * we would reach here only during assoc, so we would have all
+	 * caps to calculate the nss.
+	 */
+	if (&sta->deflink != link_sta)
+		ieee80211_sta_set_rx_nss(link_sta);
+
 	if (params->opmode_notif_used) {
 		/* returned value is only needed for rc update, but the
 		 * rc isn't initialized here yet, so ignore it
--- a/net/mac80211/eht.c
+++ b/net/mac80211/eht.c
@@ -79,7 +79,7 @@ ieee80211_eht_cap_ie_to_sta_eht_cap(stru
 	link_sta->pub->bandwidth = ieee80211_sta_cur_vht_bw(link_sta);
 }
 
-void ieee80211_eht_op_ie_to_bss_conf(struct ieee80211_vif *vif,
+void ieee80211_eht_op_ie_to_bss_conf(struct ieee80211_bss_conf *link_conf,
 				     const struct ieee80211_eht_operation *op_ie)
 {
 	struct ieee80211_eht_operation_info *eht_oper_info;
@@ -90,12 +90,12 @@ void ieee80211_eht_op_ie_to_bss_conf(str
 	if (op_ie->params && IEEE80211_EHT_OPER_INFO_PRESENT) {
 		eht_oper_info = op_ie->optional;
 
-		vif->bss_conf.eht_oper.chan_width =
+		link_conf->eht_oper.chan_width =
 				u8_get_bits(eht_oper_info->control,
 					    IEEE80211_EHT_OPER_CHAN_WIDTH);
-		vif->bss_conf.eht_oper.ccfs0 = eht_oper_info->ccfs0;
-		vif->bss_conf.eht_oper.ccfs1 = eht_oper_info->ccfs1;
-		vif->bss_conf.eht_oper.present_bm =
+		link_conf->eht_oper.ccfs0 = eht_oper_info->ccfs0;
+		link_conf->eht_oper.ccfs1 = eht_oper_info->ccfs1;
+		link_conf->eht_oper.present_bm =
 				u8_get_bits(eht_oper_info->control,
 					    IEEE80211_EHT_OPER_DISABLED_SUBCHANNEL_BITMAP_PRESENT);
 	}
--- a/net/mac80211/ieee80211_i.h
+++ b/net/mac80211/ieee80211_i.h
@@ -2600,6 +2600,6 @@ ieee80211_eht_cap_ie_to_sta_eht_cap(stru
 				    const struct ieee80211_eht_cap_elem *eht_cap_ie_elem,
 				    u8 eht_cap_len,
 				    struct link_sta_info *link_sta);
-void ieee80211_eht_op_ie_to_bss_conf(struct ieee80211_vif *vif,
+void ieee80211_eht_op_ie_to_bss_conf(struct ieee80211_bss_conf *link_conf,
 				     const struct ieee80211_eht_operation *op_ie);
 #endif /* IEEE80211_I_H */
--- a/net/mac80211/mesh.c
+++ b/net/mac80211/mesh.c
@@ -784,7 +784,7 @@ ieee80211_mesh_update_bss_params(struct
 		eht_oper = (void *)(cap->data + 1);
 
 	if (eht_oper)
-		ieee80211_eht_op_ie_to_bss_conf(&sdata->vif, eht_oper);
+		ieee80211_eht_op_ie_to_bss_conf(&sdata->vif.bss_conf, eht_oper);
 }
 
 /**
--- a/net/mac80211/agg-rx.c
+++ b/net/mac80211/agg-rx.c
@@ -207,6 +207,8 @@ static void ieee80211_send_addba_resp(st
 	struct ieee80211_mgmt *mgmt;
 	bool amsdu = ieee80211_hw_check(&local->hw, SUPPORTS_AMSDU_IN_AMPDU);
 	u16 capab;
+	struct ieee80211_link_data *link;
+	u8 link_id;
 
 	skb = dev_alloc_skb(sizeof(*mgmt) +
 		    2 + sizeof(struct ieee80211_addba_ext_ie) +
@@ -218,14 +220,29 @@ static void ieee80211_send_addba_resp(st
 	mgmt = skb_put_zero(skb, 24);
 	memcpy(mgmt->da, da, ETH_ALEN);
 	memcpy(mgmt->sa, sdata->vif.addr, ETH_ALEN);
-	if (sdata->vif.type == NL80211_IFTYPE_AP ||
-	    sdata->vif.type == NL80211_IFTYPE_AP_VLAN ||
-	    sdata->vif.type == NL80211_IFTYPE_MESH_POINT)
+	if (sdata->vif.type == NL80211_IFTYPE_AP) {
 		memcpy(mgmt->bssid, sdata->vif.addr, ETH_ALEN);
-	else if (sdata->vif.type == NL80211_IFTYPE_STATION)
+		/* Override MLD address in A2/A3 with link addr
+		 * in case of AP MLO and a non ML STA
+		 */
+		if (sdata->vif.valid_links && sta && !sta->sta.mlo) {
+			link_id = sta->deflink.link_id;
+			rcu_read_lock();
+			link = rcu_dereference(sdata->link[link_id]);
+			if (link) {
+				memcpy(mgmt->sa, link->conf->addr, ETH_ALEN);
+				memcpy(mgmt->bssid, link->conf->addr, ETH_ALEN);
+			}
+			rcu_read_unlock();
+		}
+	} else if (sdata->vif.type == NL80211_IFTYPE_AP_VLAN ||
+		sdata->vif.type == NL80211_IFTYPE_MESH_POINT) {
+		memcpy(mgmt->bssid, sdata->vif.addr, ETH_ALEN);
+	} else if (sdata->vif.type == NL80211_IFTYPE_STATION) {
 		memcpy(mgmt->bssid, sdata->deflink.u.mgd.bssid, ETH_ALEN);
-	else if (sdata->vif.type == NL80211_IFTYPE_ADHOC)
+	} else if (sdata->vif.type == NL80211_IFTYPE_ADHOC) {
 		memcpy(mgmt->bssid, sdata->u.ibss.bssid, ETH_ALEN);
+	}
 
 	mgmt->frame_control = cpu_to_le16(IEEE80211_FTYPE_MGMT |
 					  IEEE80211_STYPE_ACTION);
@@ -269,6 +286,9 @@ void ___ieee80211_start_rx_ba_session(st
 	int i, ret = -EOPNOTSUPP;
 	u16 status = WLAN_STATUS_REQUEST_DECLINED;
 	u16 max_buf_size;
+	u8 link_id;
+	enum nl80211_band band;
+	struct ieee80211_bss_conf *link_conf;
 
 	if (tid >= IEEE80211_FIRST_TSPEC_TSID) {
 		ht_dbg(sta->sdata,
@@ -277,8 +297,22 @@ void ___ieee80211_start_rx_ba_session(st
 		goto end;
 	}
 
-	if (!sta->sta.deflink.ht_cap.ht_supported &&
-	    sta->sdata->vif.bss_conf.chandef.chan->band != NL80211_BAND_6GHZ) {
+	link_id = sta->sta.deflink.link_id;
+	rcu_read_lock();
+	link_conf = rcu_dereference(sta->sdata->vif.link_conf[link_id]);
+
+	if (WARN_ON(!link_conf || (!link_conf->chandef.chan))) {
+		ht_dbg(sta->sdata,
+		       "STA %pM BA session couldnt setup due to invalid link %d\n",
+		        sta->sta.addr, link_id);
+		rcu_read_unlock();
+		goto end;
+	}
+
+	band = link_conf->chandef.chan->band;
+
+	rcu_read_unlock();
+	if (!sta->sta.deflink.ht_cap.ht_supported && band != NL80211_BAND_6GHZ) {
 		ht_dbg(sta->sdata,
 		       "STA %pM erroneously requests BA session on tid %d w/o QoS\n",
 		       sta->sta.addr, tid);
